src/protocol/handshake.rs

```diff
@@ -42,55 +43,104 @@ impl SecWebSocket {
-                req.extend_from_slice(
-                    format!("Sec-WebSocket-Version: {}\r\n", SEC_WEBSOCKET_VERSION).as_bytes(),
-                );
-                req.extend_from_slice(b"Upgrade: websocket\r\n");
-                req.extend_from_slice(
-                    format!("X-Algorand-Accept-Version: {}\r\n", X_AG_ACCEPT_VERSION).as_bytes(),
-                );
-                req.extend_from_slice(
-                    format!("X-Algorand-Instancename: {}\r\n", X_AG_INSTANCE_NAME).as_bytes(),
-                );
-                req.extend_from_slice(b"X-Algorand-Location: \r\n");
-                req.extend_from_slice(
-                    format!("X-Algorand-Noderandom: {}\r\n", X_AG_NODE_RANDOM).as_bytes(),
-                );
-                // req.extend_from_slice(b"X-Algorand-Telid: d12c01a5-4ca4-4be3-a394-68c8913f3883\r\n"); // TODO: Investigate more
-                req.extend_from_slice(
-                    format!("X-Algorand-Version: {}\r\n", X_AG_ALGORAND_VERSION).as_bytes(),
-                );
-                req.extend_from_slice(
-                    format!("X-Algorand-Genesis: {}\r\n", X_AG_ALGORAND_GENESIS).as_bytes(),
-                );
-                req.extend_from_slice(b"\r\n");
-                let req = Bytes::from(req);
+                let mut req_header = |mut header: String| {
+                    header.push_str("\r\n");
+                    req.extend_from_slice(header.as_bytes());
+                };
 
-                info!(parent: self.node().span(), "sending handshake message: {:?}", req);
+                req_header(format!("GET /v1/{}/gossip HTTP/1.1", X_AG_ALGORAND_GENESIS));
+                req_header(format!("Host: {}", conn_addr));
+                req_header(format!("User-Agent: {}", cfg.user_agent));
+                req_header("Connection: Upgrade".into());
+                req_header(format!("Sec-WebSocket-Key: {}", sec_ws.key));
+                req_header(format!("Sec-WebSocket-Version: {}", cfg.ws_version));
+                req_header("Upgrade: websocket".into());
+                req_header(format!(
+                    "X-Algorand-Accept-Version: {}",
+                    cfg.ar_accept_version
+                ));
+                req_header(format!("X-Algorand-Instancename: {}", cfg.ar_instance_name));
+                if let Some(ref location) = cfg.ar_location {
+                    req_header(format!("X-Algorand-Location: {location}"));
+                }
+                req_header(format!("X-Algorand-Noderandom: {}", cfg.ar_node_random));
+                if let Some(ref telid) = cfg.ar_tel_id {
+                    req_header(format!("X-Algorand-Telid: {telid}"));
+                }
+                req_header(format!("X-Algorand-Version: {}", cfg.ar_version));
+                req_header(format!("X-Algorand-Genesis: {}", cfg.ar_genesis));
+                req_header("".into()); // A HTTP header ends with '\r\n'
```

Comment:
Every header must end with one `\r\n`, and then after the last header, we must append one more empty `\r\n`

> req_header doesn't add one at the end of the line.

req_header does add one at the end of the line? 🕵️‍♀️